### PR TITLE
refactor(nodebuilder/share/cmd)!: Take block height instead of `ExtendedHeader` as argument for share cmds

### PR DESCRIPTION
**Pull request by bao1029p**
**Description**

Replace extended headers with block height as an argument in shares cmd

**Issues close**

Close [#2848 ](https://github.com/celestiaorg/celestia-node/issues/2848)

**Changes made**
- Replace all uses of `extended header` as an argument with block height in `nodebuilder/share/cmd/share.go.`
- Add logic to fetch header first with `block height` for

### CODE REVIEW

File: nodebuilder/share/cmd/share.go [hunk 1]
```diff
@@ -72,7 +71,7 @@ var sharesAvailableCmd = &cobra.Command{
 }
 
 var getSharesByNamespaceCmd = &cobra.Command{
-	Use:   "get-by-namespace [extended header, namespace]",
+	Use:   "get-by-namespace [block height, namespace]",
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:74]: Please also make it work with block hash as well, so its either height or hash

**Review by Wondertan — APPROVED** [review]
Thank you! 
Have one more request

File: nodebuilder/share/cmd/share.go (hunk 1 above)
Comment (bao1029p) [reply nodebuilder/share/cmd/share.go:74]: gotcha

Comment (renaynay) [reply nodebuilder/share/cmd/share.go:74]: @Wondertan as the issue described just using height for now, let's merge this as is. Eventually we can add support for hash or height IMO

Comment (Wondertan) [reply nodebuilder/share/cmd/share.go:74]: It's trivial to add and I don't see why we wouldn't while we are on it

Comment (Wondertan) [reply nodebuilder/share/cmd/share.go:74]: @bao1029p, could you give us an estimate and tell if you will be able to finish this until Monday(included)?

Comment (bao1029p) [reply nodebuilder/share/cmd/share.go:74]: i just done with this, will push it now

**Comment by bao1029p** [comment]
hey @Wondertan, i added both usage of block hash and height in

File: nodebuilder/share/cmd/share.go [hunk 2]
```diff
@@ -39,13 +40,18 @@ var sharesAvailableCmd = &cobra.Command{
...
 		defer client.Close()
 
-		raw, err := parseJSON(args[0])
+		var eh *header.ExtendedHeader
+		hash, err := hex.DecodeString(args[0])
 		if err != nil {
-			return err
+			height, err := strconv.ParseUint(args[0], 10, 64)
+			if err != nil {
+				return fmt.Errorf("error processing a height/hash:%v", err)
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:48]:
```suggestion
				return fmt.Errorf("can't parse the height/hash argument: %w", err)
```

File: nodebuilder/share/cmd/share.go [hunk 3]
```diff
@@ -72,7 +78,7 @@ var sharesAvailableCmd = &cobra.Command{
 }
 
 var getSharesByNamespaceCmd = &cobra.Command{
-	Use:   "get-by-namespace [extended header, namespace]",
+	Use:   "get-by-namespace [block height or hash, namespace]",
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:81]:
```suggestion
	Use:   "get-by-namespace (height | hash) namespace]",
```

File: nodebuilder/share/cmd/share.go [hunk 4]
```diff
@@ -158,8 +174,8 @@ var getShare = &cobra.Command{
 }
 
 var getEDS = &cobra.Command{
-	Use:   "get-eds [extended header]",
-	Short: "Gets the full EDS identified by the given extended header",
+	Use:   "get-eds [block height or hash]",
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:177]:
```suggestion
	Use:   "get-eds (height | hash)",
```

The usage of square brackets was wrong. It should be used to show optionality while here the arguments is mandatory. Also, for mutually exclusive arguments | must be used.

Comment (Wondertan) [reply nodebuilder/share/cmd/share.go:177]: See https://pubs.opengroup.org/onlinepubs/000095399/basedefs/xbd_chap12.html#tag_12_01

File: nodebuilder/share/cmd/share.go [hunk 5]
```diff
@@ -104,7 +115,7 @@ var getSharesByNamespaceCmd = &cobra.Command{
 }
 
 var getShare = &cobra.Command{
-	Use:   "get-share [extended header, row, col]",
+	Use:   "get-share [block height or hash, row, col]",
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:118]:
```suggestion
	Use:   "get-share (height | hash) row col",
```

File: nodebuilder/share/cmd/share.go [hunk 6]
```diff
@@ -82,13 +88,18 @@ var getSharesByNamespaceCmd = &cobra.Command{
...
 		defer client.Close()
 
-		raw, err := parseJSON(args[0])
+		var eh *header.ExtendedHeader
+		hash, err := hex.DecodeString(args[0])
 		if err != nil {
-			return err
+			height, err := strconv.ParseUint(args[0], 10, 64)
+			if err != nil {
+				return fmt.Errorf("error processing a height/hash:%v", err)
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:96]:
```suggestion
				return fmt.Errorf("can't parse the height/hash argument: %w", err)
```

File: nodebuilder/share/cmd/share.go [hunk 7]
```diff
@@ -114,13 +125,18 @@ var getShare = &cobra.Command{
...
 		defer client.Close()
 
-		raw, err := parseJSON(args[0])
+		var eh *header.ExtendedHeader
+		hash, err := hex.DecodeString(args[0])
 		if err != nil {
-			return err
+			height, err := strconv.ParseUint(args[0], 10, 64)
+			if err != nil {
+				return fmt.Errorf("error parsing a height/hash:%v", err)
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:133]:
```suggestion
				return fmt.Errorf("can't parse the height/hash argument: %w", err)
```

File: nodebuilder/share/cmd/share.go [hunk 8]
```diff
@@ -168,13 +184,18 @@ var getEDS = &cobra.Command{
...
 		defer client.Close()
 
-		raw, err := parseJSON(args[0])
+		var eh *header.ExtendedHeader
+		hash, err := hex.DecodeString(args[0])
 		if err != nil {
-			return err
+			height, err := strconv.ParseUint(args[0], 10, 64)
+			if err != nil {
+				return fmt.Errorf("error processing a height/hash:%v", err)
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:192]:
```suggestion
				return fmt.Errorf("can't parse the height/hash argument: %w", err)
```

Comment (Wondertan) [reply nodebuilder/share/cmd/share.go:192]: you may wann to extract that in a separate func as its repeated 4 times.

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: nodebuilder/share/cmd/share.go [hunk 9]
```diff
@@ -184,8 +166,18 @@ var getEDS = &cobra.Command{
...
+		if err != nil {
+			return nil, fmt.Errorf("can't parse the height/hash argument: %w", err)
+		}
+		eh, err = client.Header.GetByHeight(ctx, height)
+	} else {
+		eh, err = client.Header.GetByHash(ctx, hash)
+	}
+
+	return eh, err
+}
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:183]: Needs new line

File: nodebuilder/share/cmd/share.go [hunk 10]
```diff
@@ -104,7 +96,7 @@ var getSharesByNamespaceCmd = &cobra.Command{
 }
 
 var getShare = &cobra.Command{
-	Use:   "get-share [extended header, row, col]",
+	Use:   "get-share [(height | hash), row, col]",
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:99]:
```suggestion
	Use:   "get-share (height | hash) row col",
```

File: nodebuilder/share/cmd/share.go [hunk 11]
```diff
@@ -72,7 +69,7 @@ var sharesAvailableCmd = &cobra.Command{
 }
 
 var getSharesByNamespaceCmd = &cobra.Command{
-	Use:   "get-by-namespace [extended header, namespace]",
+	Use:   "get-by-namespace [(height | hash), namespace]",
```
Comment (Wondertan) [inline nodebuilder/share/cmd/share.go:72]:
```suggestion
	Use:   "get-by-namespace (height | hash) namespace",
```

File: nodebuilder/share/cmd/share.go [hunk 12]
```diff
@@ -184,8 +166,18 @@ var getEDS = &cobra.Command{
...
+		height, err := strconv.ParseUint(arg, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("can't parse the height/hash argument: %w", err)
+		}
+		eh, err = client.Header.GetByHeight(ctx, height)
+	} else {
+		eh, err = client.Header.GetByHash(ctx, hash)
+	}
+
+	return eh, err
```
Comment (vgonkivs) [inline nodebuilder/share/cmd/share.go:182]:
```suggestion
	hash, err := hex.DecodeString(arg)
	if err == nil {
		return client.Header.GetByHash(ctx, hash)
	}

	height, err := strconv.ParseUint(arg, 10, 64)
	if err != nil {
		return nil, fmt.Errorf("can't parse the height/hash argument: %w", err)
	}
	return client.Header.GetByHeight(ctx, height)
```

**Review by vgonkivs — CHANGES_REQUESTED** [review]

**Review by Wondertan — APPROVED** [review]
Approving. Did you test it out manually? Please do if not and then we can merge

**Comment by bao1029p** [comment]
> Approving. Did you test it out manually? Please do if not and then we can merge

yes i've tested it and everything work great

**Review by vgonkivs — APPROVED** [review]

**Review by distractedm1nd — APPROVED** [review]
utACK

**Review by renaynay — APPROVED** [review]

**Pull request merged**